### PR TITLE
Custom vl user input

### PR DESCRIPTION
**Pull request by irexyc**
## Motivation

for some models, the location of the image token is not restricted, in this case, the user should insert the image token manually.

### CODE REVIEW

File: docs/en/inference/vl_pipeline.md [hunk 1]
```diff
@@ -97,6 +97,22 @@ response = pipe(('describe this image', image), gen_config=gen_config)
 print(response)
 ```
 
+### Set image token position
```
Comment (lvhan028) [inline docs/en/inference/vl_pipeline.md:100]: Customize image token positioni

File: docs/en/inference/vl_pipeline.md [hunk 2]
```diff
@@ -97,6 +97,22 @@ response = pipe(('describe this image', image), gen_config=gen_config)
...
+
+By default, LMDeploy will insert the special image token into the user prompt based on the chat template provided by the algorithm repo. However, for some models, the location of the image token is not restricted, such as deepseek-vl, or the user needs to customize the image token position. In these cases, the user needs to manually insert the special image token into the prompt.
```
Comment (lvhan028) [inline docs/en/inference/vl_pipeline.md:102]: "will insert" -> "inserts"

File: lmdeploy/vl/templates.py [hunk 3]
```diff
@@ -136,6 +139,10 @@ class DeepSeekVLChatTemplateWrapper(VLChatTemplateWrapper):
 
     def append_image_token(self, prompt, num_images: int):
         """append image tokens to user prompt."""
+        logger.error(
+            f'for deepseek-vl model, the user should insert the {IMAGE_TOKEN} '
+            'to user prompt manually, please read '
+            'docs/en/inference/vl_pipeline.md for more details.')
```
Comment (lvhan028) [inline lmdeploy/vl/templates.py:145]: Let's provide the complete url of the user guide

File: lmdeploy/serve/vl_async_engine.py [hunk 4]
```diff
@@ -57,8 +60,13 @@ async def _get_prompt_input(self, prompt: Dict, do_preprocess: bool,
             input_ids = []
             begins = []
             ends = []
+            if len(segs) != len(features) + 1:
+                logger.error(
+                    f'the number of {IMAGE_TOKEN} is not equal '
+                    f'to input images, {len(segs) - 1} vs {len(features)}')
+                features = features[:len(segs) - 1]
```
Comment (lvhan028) [inline lmdeploy/serve/vl_async_engine.py:67]: Is it possible that `len(segs) > len(features)`

File: docs/zh_cn/inference/vl_pipeline.md [hunk 5]
```diff
@@ -97,6 +97,22 @@ response = pipe(('describe this image', image), gen_config=gen_config)
...
+### 自定义图片 token 的位置
+
+默认情况下，LMDeploy 会根据算法 repo 提供的对话模版将表示图片的特殊 token 插入到 user prompt 中，但在一些模型中，图片 token 的位置并没有限制，如 deepseek-vl，或者用户需要自定义图片 token 插入的位置。这种情况下，用户需要手动将表示图片的 token 插入到 prompt 中。
+
+```python
+from lmdeploy import pipeline
+from lmdeploy.vl import load_image
+from lmdeploy.vl.constants import IMAGE_TOKEN
+
+pipe = pipeline('deepseek-ai/deepseek-vl-7b-chat')
```
Comment (AllentDan) [inline docs/zh_cn/inference/vl_pipeline.md:109]: Shall we try `deepseek-ai/deepseek-vl-1.3b-chat` instead as an example? It is smaller and can be downloaded faster.

File: docs/zh_cn/inference/vl_pipeline.md [hunk 6]
```diff
@@ -97,6 +97,22 @@ response = pipe(('describe this image', image), gen_config=gen_config)
...
 ```
 
+### 自定义图片 token 的位置
+
+默认情况下，LMDeploy 会根据算法 repo 提供的对话模版将表示图片的特殊 token 插入到 user prompt 中，但在一些模型中，图片 token 的位置并没有限制，如 deepseek-vl，或者用户需要自定义图片 token 插入的位置。这种情况下，用户需要手动将表示图片的 token 插入到 prompt 中。
+
+```python
+from lmdeploy import pipeline
+from lmdeploy.vl import load_image
+from lmdeploy.vl.constants import IMAGE_TOKEN
```
Comment (AllentDan) [inline docs/zh_cn/inference/vl_pipeline.md:107]: May notify users that is is actually `'<IMAGE_TOKEN>'` in comments or other forms. This is useful for api_server users.

**Review by AllentDan — APPROVED** [review]
LGTM

**Pull request merged**